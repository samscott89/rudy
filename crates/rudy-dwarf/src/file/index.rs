use std::{collections::BTreeSet, path::PathBuf};

use gimli::Reader as _;

use crate::{
    die::{
        cu::is_rust_cu, file_entry_to_path, navigation::get_roots, utils::pretty_print_die_entry,
    },
    DebugFile, DwarfDb, SourceFile,
};

/// Build an index of all types + functions (using fully qualified names
/// that can be extracted from demangled symbols) to their
/// corresponding DIE entry in the DWARF information.
#[salsa::tracked(returns(ref))]
pub fn index_debug_file_sources<'db>(
    db: &'db dyn DwarfDb,
    debug_file: DebugFile,
) -> (BTreeSet<PathBuf>, BTreeSet<SourceFile<'db>>) {
    let mut compile_dirs = BTreeSet::new();
    let mut sources = BTreeSet::new();

    let roots = get_roots(db, debug_file);
    for (_unit_offset, unit_ref) in &roots {
        let mut entries = unit_ref.entries();
        let Some((_, root)) = entries.next_dfs().ok().flatten() else {
            continue;
        };
        if is_rust_cu(root, unit_ref) {
            // get the compile directory
            if let Some(compile_dir) = &unit_ref.comp_dir {
                match compile_dir.to_string() {
                    Ok(compile_dir) => {
                        // if the compile directory is empty, we can skip it
                        if compile_dir.is_empty() {
                            tracing::debug!(
                                "Skipping empty compile directory for unit: {}",
                                pretty_print_die_entry(root, unit_ref)
                            );
                        } else {
                            compile_dirs.insert(PathBuf::from(compile_dir.to_string()));
                        }
                    }
                    Err(e) => {
                        tracing::error!("Failed to convert compile directory to string: {e}");
                    }
                }
            }

            // get all referenced files
            let files = unit_ref
                .line_program
                .as_ref()
                .map(|lp| {
                    lp.header()
                        .file_names()
                        .iter()
                        .flat_map(|f| {
                            file_entry_to_path(db, f, unit_ref)
                                .map(|path| SourceFile::new(db, path))
                        })
                        .collect::<BTreeSet<_>>()
                })
                .unwrap_or_default();
            sources.extend(files);
        }
    }

    (compile_dirs, sources)
}
