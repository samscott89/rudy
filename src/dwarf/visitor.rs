use super::loader::{DwarfReader, RawDie};
use super::unit::UnitRef;
use super::{CompilationUnitId, Die};
use crate::database::Db;
use crate::file::File;
use gimli::DebuggingInformationEntry;

/// Control flow for visitor traversal
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum VisitorControl {
    /// Continue traversing the tree (visit children)
    Continue,
    /// Skip children of current node, continue with siblings
    SkipChildren,
    /// Stop traversal entirely
    Break,
}

/// Walker that drives the visitor through the DIE tree
pub struct DieWalker<'a, 'db, V> {
    pub db: &'db dyn Db,
    pub visitor: &'a mut V,

    /// The file being walked
    file: File,
    /// The compilation unit being walked
    unit_offset: gimli::UnitSectionOffset<usize>,
    pub unit_ref: &'a UnitRef<'a>,
    /// The entries cursor for the current unit
    // unit_ref: &'a UnitRef<'a>,
    cursor: gimli::EntriesCursor<'a, 'a, DwarfReader>,
}

pub fn walk_file<'db, 'a, V: DieVisitor<'db>>(db: &'db dyn Db, file: File, visitor: &'a mut V) {
    // Get the root compilation units for this file
    let roots = super::navigation::get_roots(db, file);

    for (unit_offset, unit_ref) in &roots {
        // Create a walker for each unit
        let mut walker = DieWalker {
            db,
            visitor: &mut *visitor,
            file,
            unit_offset: *unit_offset,
            unit_ref,
            cursor: unit_ref.entries(),
        };

        // Walk the compilation unit
        walker.walk_unit();
    }
}

impl<'a, 'db, V: DieVisitor<'db>> DieWalker<'a, 'db, V> {
    pub fn get_die(&self, raw: RawDie<'a>) -> Die<'db> {
        Die::new(self.db, self.file, self.unit_offset, raw.offset())
    }

    fn next_dfs(&mut self) -> Option<(isize, RawDie<'a>)> {
        match self.cursor.next_dfs() {
            Ok(Some((offset, die))) => Some((offset, die.clone())),
            Ok(None) => None,
            Err(e) => {
                self.db.report_error(format!("Failed to parse DIE: {e}"));
                None
            }
        }
    }

    fn next_sibling(&mut self) -> Option<RawDie<'a>> {
        match self.cursor.next_sibling() {
            Ok(Some(die)) => Some(die.clone()),
            Ok(None) => None,
            Err(e) => {
                self.db.report_error(format!("Failed to parse DIE: {e}"));
                None
            }
        }
    }

    pub fn walk_unit(&mut self) {
        let Some((delta, root)) = self.next_dfs() else {
            // empty tree -- nothing to walk
            self.db
                .report_info("No entries found in DWARF tree".to_string());
            return;
        };

        debug_assert!(delta == 0, "Expected root DIE to be at depth 0");

        // first entry _should_ be the root DIE -- the compilation unit
        let tag = root.tag();
        if tag != gimli::DW_TAG_compile_unit {
            self.db.report_error(format!(
                "Expected root DIE to be a compilation unit, found: {tag:?}"
            ));
            return;
        }

        let unit_ref = self.unit_ref.clone();
        V::visit_cu(self, root, unit_ref);
    }

    fn walk_generic_children(&mut self) {
        // walk the children of the compilation unit
        if let Some((delta, first_child)) = self.next_dfs() {
            debug_assert_eq!(delta, 1, "Expected first child to be at depth 1");
            V::visit_die(self, first_child, self.unit_ref.clone());

            while let Some(next) = self.next_sibling() {
                // continue walking siblings
                V::visit_die(self, next, self.unit_ref.clone());
            }
        }
    }

    pub fn walk_cu(&mut self) {
        self.walk_generic_children();
    }

    pub fn walk_namespace(&mut self) {
        self.walk_generic_children();
    }

    // pub fn walk_namespace(&mut self, die: RawDie<'a>, unit_ref: UnitRef<'a>) {
    //     if let Some(mut tree) = self
    //         .unit_ref
    //         .entries_tree(Some(die.die_offset()))
    //         .inspect_err(|e| {
    //             self.db
    //                 .report_critical(format!("Failed to parse child nodes: {e}"));
    //         })
    //         .ok()
    //     {
    //         let Some(tree_root) = tree
    //             .root()
    //             .inspect_err(|e| {
    //                 self.db
    //                     .report_critical(format!("Failed to parse child nodes: {e}"));
    //             })
    //             .ok()
    //         else {
    //             return;
    //         };

    //         for child in tree_root.children() {
    //             let child_die = RawDie::new(self.db, die.file(), die.cu_offset(), child.offset());
    //             self.visitor.visit_namespace(self.db, &child_die, &child);
    //         }
    //     }

    //     // let Some(mut tree) = self
    //     //     .unit_ref
    //     //     .entries_tree(Some(self.die_offset(db)))
    //     //     .inspect_err(|e| {
    //     //         db.report_critical(format!("Failed to parse child nodes: {e}"));
    //     //     })
    //     //     .ok()
    //     // else {
    //     //     return children;
    //     // };
    // }

    // /// Walk a compilation unit with the given visitor
    // pub fn walk_unit(
    //     &self,
    //     unit: CompilationUnitId<'a>,
    //     visitor: &mut V,
    // ) -> Result<(), gimli::Error> {
    //     let unit_ref = unit.unit_ref(self.db).ok_or(gimli::Error::InvalidInput)?;
    //     let file = unit.file(self.db);
    //     let cu_offset = unit.offset(self.db);

    //     let mut entries = unit_ref.entries();
    //     self.walk_entries(&mut entries, &unit_ref, file, cu_offset, visitor, 0)
    // }

    // /// Internal recursive walker
    // fn walk_entries(
    //     &self,
    //     entries: &mut gimli::EntriesCursor<'_>,
    //     unit_ref: UnitRef<'_>,
    //     file: crate::file::File,
    //     cu_offset: gimli::DebugInfoOffset,
    //     visitor: &mut V,
    //     depth: usize,
    // ) -> Result<(), gimli::Error> {
    //     loop {
    //         let Some((_, entry)) = entries.next_dfs()? else {
    //             break;
    //         };

    //         // Create Die handle
    //         let die = Die::new(self.db, file, cu_offset, entry.offset());

    //         // Visit the DIE
    //         let control = visitor.visit_die(self.db, &die, &entry);

    //         match control {
    //             VisitorControl::Continue => {
    //                 if entry.has_children() {
    //                     visitor.enter_scope(self.db, &die, &entry);
    //                     // Continue traversal - next_dfs will handle children
    //                 }
    //             }
    //             VisitorControl::SkipChildren => {
    //                 // Skip to next sibling
    //                 entries.skip_attributes()?;
    //                 while entries.current().is_some() && entries.current().unwrap().1 > depth {
    //                     entries.next_dfs()?;
    //                 }
    //             }
    //             VisitorControl::Break => {
    //                 return Ok(());
    //             }
    //         }

    //         // Check if we're leaving a scope
    //         if let Some((next_depth, _)) = entries.current() {
    //             if next_depth < depth && depth > 0 {
    //                 // We're about to leave this scope
    //                 visitor.leave_scope(self.db, &die, &entry);
    //             }
    //         }
    //     }

    //     Ok(())
    // }
}

/// Visitor trait for walking DWARF DIE trees
pub trait DieVisitor<'db>: Sized {
    fn visit_cu<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        _die: RawDie<'a>,
        _unit_ref: UnitRef<'a>,
    ) {
        walker.walk_cu();
    }

    /// Called for each DIE entry
    fn visit_die<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        die: RawDie<'a>,
        unit_ref: UnitRef<'a>,
    ) {
        // Default: dispatch to specific visit methods based on tag
        match die.tag() {
            gimli::DW_TAG_namespace => Self::visit_namespace(walker, die, unit_ref),
            gimli::DW_TAG_subprogram => Self::visit_function(walker, die, unit_ref),
            gimli::DW_TAG_structure_type => Self::visit_struct(walker, die, unit_ref),
            gimli::DW_TAG_enumeration_type => Self::visit_enum(walker, die, unit_ref),
            gimli::DW_TAG_variable => Self::visit_variable(walker, die, unit_ref),
            gimli::DW_TAG_base_type => Self::visit_base_type(walker, die, unit_ref),
            _ => {}
        }
    }

    /// Visit a namespace
    fn visit_namespace<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        _entry: RawDie<'a>,
        _unit_ref: UnitRef<'a>,
    ) {
        walker.walk_namespace();
    }

    /// Visit a function/subprogram
    fn visit_function<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        _entry: RawDie<'a>,
        unit_ref: UnitRef<'a>,
    ) {
    }

    /// Visit a struct type
    fn visit_struct<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        _entry: RawDie<'a>,
        unit_ref: UnitRef<'a>,
    ) {
    }

    /// Visit an enum type
    fn visit_enum<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        _entry: RawDie<'a>,
        unit_ref: UnitRef<'a>,
    ) {
    }

    /// Visit a variable
    fn visit_variable<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        _entry: RawDie<'a>,
        unit_ref: UnitRef<'a>,
    ) {
    }

    /// Visit a base type
    fn visit_base_type<'a>(
        walker: &mut DieWalker<'a, 'db, Self>,
        _entry: RawDie<'a>,
        unit_ref: UnitRef<'a>,
    ) {
    }

    // /// Called when entering a new scope (before visiting children)
    // fn enter_scope(walker: &mut DieWalker<'a, 'db, Self>, _entry: RawDie<'a>, unit_ref: UnitRef<'a>) {}

    // /// Called when leaving a scope (after visiting children)
    // fn leave_scope(walker: &mut DieWalker<'a, 'db, Self>, _entry: RawDie<'a>, unit_ref: UnitRef<'a>) {}
}
