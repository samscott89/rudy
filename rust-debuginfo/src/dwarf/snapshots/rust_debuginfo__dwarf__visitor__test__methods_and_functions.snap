---
source: src/dwarf/visitor.rs
expression: "visitor.functions.join(\"\\n\")"
snapshot_kind: text
---
std::rt::lang_start::{closure#0}<()>
std::rt::lang_start<()>
std::hash::random::new
std::hash::random::{impl#5}::default
std::hash::random::{impl#5}::default
std::hash::random::{impl#0}::new::{closure#0}
std::hash::random::{impl#0}::new::KEYS::{constant#0}::{closure#1}
std::hash::random::{impl#0}::new::KEYS::__init
std::collections::hash::map::{impl#7}::default<alloc::string::String, i32, std::hash::random::RandomState>
std::collections::hash::map::{impl#7}::default<alloc::string::String, small::TestStruct0, std::hash::random::RandomState>
std::collections::hash::map::with_hasher<alloc::string::String, i32, std::hash::random::RandomState>
std::collections::hash::map::new<alloc::string::String, i32>
std::collections::hash::map::with_hasher<alloc::string::String, small::TestStruct0, std::hash::random::RandomState>
std::collections::hash::map::new<alloc::string::String, small::TestStruct0>
std::sys::process::unix::common::as_i32
std::sys::thread_local::native::lazy::initialize<core::cell::Cell<(u64, u64)>, !, fn() -> core::cell::Cell<(u64, u64)>>
std::sys::thread_local::native::lazy::get_or_init<core::cell::Cell<(u64, u64)>, !, fn() -> core::cell::Cell<(u64, u64)>>
std::sys::thread_local::native::lazy::{impl#0}::register_dtor<core::cell::Cell<(u64, u64)>>
std::sys::backtrace::__rust_begin_short_backtrace<fn(), ()>
std::process::to_i32
std::process::{impl#59}::report
std::thread::local::with<core::cell::Cell<(u64, u64)>, std::hash::random::{impl#0}::new::{closure_env#0}, std::hash::random::RandomState>
std::thread::local::try_with<core::cell::Cell<(u64, u64)>, std::hash::random::{impl#0}::new::{closure_env#0}, std::hash::random::RandomState>
core::ptr::unique::new_unchecked<u8>
core::ptr::non_null::without_provenance<u8>
core::ptr::non_null::as_ptr<u8>
core::ptr::non_null::new<u8>
core::ptr::non_null::new_unchecked<u8>
core::ptr::non_null::cast<u8, (alloc::string::String, i32)>
core::ptr::non_null::cast<u8, (alloc::string::String, small::TestStruct0)>
core::ptr::non_null::{impl#3}::new_unchecked::precondition_check
core::ptr::non_null::slice_from_raw_parts<u8>
core::ptr::non_null::new_unchecked<[u8]>
core::ptr::non_null::as_ptr<(alloc::string::String, i32)>
core::ptr::non_null::as_ptr<(alloc::string::String, small::TestStruct0)>
core::ptr::replace<std::sys::thread_local::native::lazy::State<core::cell::Cell<(u64, u64)>, !>>
core::ptr::mut_ptr::{impl#0}::replace<std::sys::thread_local::native::lazy::State<core::cell::Cell<(u64, u64)>, !>>
core::ptr::mut_ptr::{impl#0}::cast<u8, u8>
core::ptr::mut_ptr::{impl#0}::sub<(alloc::string::String, i32)>
core::ptr::mut_ptr::{impl#0}::drop_in_place<(alloc::string::String, i32)>
core::ptr::mut_ptr::{impl#0}::sub<(alloc::string::String, small::TestStruct0)>
core::ptr::mut_ptr::{impl#0}::drop_in_place<(alloc::string::String, small::TestStruct0)>
core::ptr::mut_ptr::{impl#0}::sub<u8>
core::ptr::mut_ptr::{impl#0}::sub<u8>
core::ptr::mut_ptr::{impl#0}::sub<(alloc::string::String, i32)>
core::ptr::mut_ptr::{impl#0}::sub<(alloc::string::String, small::TestStruct0)>
core::ptr::metadata::from_raw_parts<core::cell::Cell<(u64, u64)>, ()>
core::ptr::metadata::from_raw_parts_mut<[u8], u8>
core::ptr::null<core::cell::Cell<(u64, u64)>>
core::ptr::const_ptr::{impl#0}::as_ref<core::cell::Cell<(u64, u64)>>
core::ptr::const_ptr::{impl#0}::is_aligned_to<()>
core::ptr::const_ptr::{impl#0}::addr<()>
core::ptr::const_ptr::{impl#0}::is_null::runtime
core::ptr::const_ptr::{impl#0}::is_null<()>
core::ptr::const_ptr::{impl#0}::is_aligned_to<()>
core::ptr::const_ptr::{impl#0}::is_aligned_to<()>
core::ptr::const_ptr::{impl#0}::addr<()>
core::ptr::const_ptr::{impl#0}::copy_to_nonoverlapping<u8>
core::ptr::const_ptr::{impl#0}::align_offset<hashbrown::control::tag::Tag>
core::ptr::const_ptr::{impl#0}::add<u8>
core::ptr::const_ptr::{impl#0}::cast<u8, hashbrown::control::tag::Tag>
core::ptr::const_ptr::{impl#0}::align_offset<hashbrown::control::tag::Tag>
core::ptr::const_ptr::{impl#0}::add<u8>
core::ptr::const_ptr::{impl#0}::cast<u8, hashbrown::control::tag::Tag>
core::ptr::const_ptr::{impl#0}::align_offset<hashbrown::control::tag::Tag>
core::ptr::const_ptr::{impl#0}::add<u8>
core::ptr::const_ptr::{impl#0}::cast<u8, hashbrown::control::tag::Tag>
core::ptr::const_ptr::{impl#0}::align_offset<hashbrown::control::tag::Tag>
core::ptr::const_ptr::{impl#0}::add<u8>
core::ptr::const_ptr::{impl#0}::cast<u8, hashbrown::control::tag::Tag>
core::ptr::drop_in_place<hashbrown::raw::RawTable<(alloc::string::String, small::TestStruct0), alloc::alloc::Global>>
core::ptr::drop_in_place<std::collections::hash::map::HashMap<alloc::string::String, small::TestStruct0, std::hash::random::RandomState>>
core::ptr::drop_in_place<hashbrown::map::HashMap<alloc::string::String, i32, std::hash::random::RandomState, alloc::alloc::Global>>
core::ptr::drop_in_place<hashbrown::map::HashMap<alloc::string::String, small::TestStruct0, std::hash::random::RandomState, alloc::alloc::Global>>
core::ptr::read_volatile::precondition_check
core::ptr::read_unaligned<core::core_arch::arm_shared::neon::uint8x8_t>
core::ptr::drop_in_place<small::TestEnum0>
core::ptr::drop_in_place<small::TestStruct0>
core::ptr::drop_in_place<small::TestStruct1>
core::ptr::drop_in_place<small::TestStruct2>
core::ptr::drop_in_place<small::TestStruct3>
core::ptr::drop_in_place<small::TestStruct4>
core::ptr::drop_in_place<small::TestStruct5>
core::ptr::drop_in_place<small::TestStruct6>
core::ptr::drop_in_place<small::TestStruct7>
core::ptr::drop_in_place<small::TestStruct8>
core::ptr::drop_in_place<small::TestStruct9>
core::ptr::drop_in_place<alloc::string::String>
core::ptr::drop_in_place<alloc::vec::Vec<u8, alloc::alloc::Global>>
core::ptr::drop_in_place<alloc::raw_vec::RawVec<u8, alloc::alloc::Global>>
core::ptr::drop_in_place<(alloc::string::String, i32)>
core::ptr::drop_in_place<(alloc::string::String, small::TestStruct0)>
core::ptr::replace::precondition_check
core::ptr::drop_in_place<std::rt::lang_start::{closure_env#0}<()>>
core::ptr::drop_in_place<hashbrown::raw::RawTable<(alloc::string::String, i32), alloc::alloc::Global>>
core::ptr::drop_in_place<std::collections::hash::map::HashMap<alloc::string::String, i32, std::hash::random::RandomState>>
core::ptr::read_volatile<u8>
core::ptr::read_volatile<u8>
core::ptr::slice_from_raw_parts_mut<u8>
core::ptr::without_provenance<u8>
core::num::niche_types::new_unchecked
core::num::niche_types::as_inner
core::num::{impl#9}::wrapping_add
core::num::{impl#2}::overflowing_add
core::num::{impl#2}::overflowing_add_unsigned
core::num::{impl#2}::checked_add_unsigned
core::num::{impl#2}::wrapping_add
core::num::{impl#2}::wrapping_mul
core::num::{impl#11}::count_ones
core::num::{impl#11}::is_power_of_two
core::num::{impl#11}::count_ones
core::num::{impl#11}::is_power_of_two
core::num::{impl#11}::count_ones
core::num::{impl#11}::is_power_of_two
core::num::{impl#11}::overflowing_mul
core::num::{impl#11}::checked_mul
core::num::{impl#11}::abs_diff
core::num::{impl#11}::wrapping_sub
core::num::{impl#11}::overflowing_mul
core::num::{impl#11}::checked_mul
core::num::{impl#11}::checked_add
core::num::{impl#11}::is_power_of_two
core::num::{impl#11}::count_ones
core::num::{impl#11}::is_power_of_two
core::num::{impl#11}::count_ones
core::num::{impl#11}::is_power_of_two
core::num::{impl#11}::count_ones
core::num::{impl#11}::is_power_of_two
core::num::{impl#11}::count_ones
core::num::nonzero::new<u64>
core::num::nonzero::trailing_zeros
core::option::and_then<&mut core::option::Option<core::cell::Cell<(u64, u64)>>, core::cell::Cell<(u64, u64)>, fn(&mut core::option::Option<core::cell::Cell<(u64, u64)>>) -> core::option::Option<core::cell::Cell<(u64, u64)>>>
core::option::take<core::cell::Cell<(u64, u64)>>
core::option::unwrap_or_else<core::cell::Cell<(u64, u64)>, fn() -> core::cell::Cell<(u64, u64)>>
core::option::ok_or<&core::cell::Cell<(u64, u64)>, std::thread::local::AccessError>
core::option::unwrap_unchecked<i32>
core::option::ok_or<core::ptr::non_null::NonNull<u8>, core::alloc::AllocError>
core::option::{impl#40}::branch<usize>
core::option::{impl#40}::branch<usize>
core::option::{impl#40}::branch<usize>
core::option::{impl#41}::from_residual<(core::alloc::layout::Layout, usize)>
core::cell::get<(u64, u64)>
core::cell::replace<(u64, u64)>
core::cell::set<(u64, u64)>
core::cell::get<std::sys::thread_local::native::lazy::State<core::cell::Cell<(u64, u64)>, !>>
core::ops::function::FnOnce::call_once<fn(&mut core::option::Option<core::cell::Cell<(u64, u64)>>) -> core::option::Option<core::cell::Cell<(u64, u64)>>, (&mut core::option::Option<core::cell::Cell<(u64, u64)>>)>
core::ops::function::FnOnce::call_once<std::rt::lang_start::{closure_env#0}<()>, ()>
core::ops::function::FnOnce::call_once<fn(), ()>
core::ops::function::FnOnce::call_once<fn() -> core::cell::Cell<(u64, u64)>, ()>
core::ops::function::FnOnce::call_once<std::hash::random::{impl#0}::new::KEYS::{constant#0}::{closure_env#1}, (core::option::Option<&mut core::option::Option<core::cell::Cell<(u64, u64)>>>)>
core::ops::function::FnOnce::call_once<std::rt::lang_start::{closure_env#0}<()>, ()>
core::mem::replace<core::option::Option<core::cell::Cell<(u64, u64)>>>
core::mem::replace<std::sys::thread_local::native::lazy::State<core::cell::Cell<(u64, u64)>, !>>
core::mem::drop<std::sys::thread_local::native::lazy::State<core::cell::Cell<(u64, u64)>, !>>
core::mem::replace<(u64, u64)>
core::mem::size_of<core::core_arch::arm_shared::neon::uint8x8_t>
core::mem::manually_drop::into_inner<core::core_arch::arm_shared::neon::uint8x8_t>
core::hint::unreachable_unchecked
core::hint::black_box<()>
core::hint::unreachable_unchecked
core::hint::assert_unchecked::precondition_check
core::hint::unreachable_unchecked::precondition_check
core::hint::assert_unchecked
core::result::{impl#28}::from_residual<std::hash::random::RandomState, std::thread::local::AccessError, std::thread::local::AccessError>
core::result::{impl#28}::from_residual<core::ptr::non_null::NonNull<[u8]>, core::alloc::AllocError, core::alloc::AllocError>
core::result::{impl#27}::branch<&core::cell::Cell<(u64, u64)>, std::thread::local::AccessError>
core::result::{impl#27}::branch<core::ptr::non_null::NonNull<u8>, core::alloc::AllocError>
core::intrinsics::unlikely
core::intrinsics::copy_nonoverlapping::precondition_check
core::intrinsics::cold_path
core::intrinsics::copy_nonoverlapping<u8>
core::intrinsics::unlikely
core::intrinsics::copy_nonoverlapping<u8>
core::intrinsics::unlikely
core::iter::range::{impl#40}::forward_unchecked
core::iter::range::{impl#6}::next<i32>
core::iter::range::{impl#5}::spec_next<i32>
core::iter::traits::collect::{impl#0}::into_iter<core::ops::range::Range<i32>>
core::fmt::new_const<1>
core::fmt::new_v1<2, 1>
core::fmt::rt::new<u64>
core::fmt::rt::new_display<u64>
core::ub_checks::maybe_is_aligned_and_not_null::runtime
core::ub_checks::maybe_is_aligned_and_not_null::runtime
core::ub_checks::maybe_is_aligned_and_not_null::runtime
core::ub_checks::maybe_is_aligned_and_not_null
core::ub_checks::maybe_is_nonoverlapping
core::ub_checks::maybe_is_aligned_and_not_null
core::ub_checks::maybe_is_aligned_and_not_null
core::ub_checks::maybe_is_nonoverlapping::runtime
core::core_arch::arm_shared::neon::vget_lane_u64<0>
core::core_arch::arm_shared::neon::generated::vreinterpret_s8_u8
core::core_arch::arm_shared::neon::generated::vreinterpret_u64_u8
core::core_arch::aarch64::neon::generated::vld1_u8
core::core_arch::aarch64::neon::generated::vcgez_s8
core::core_arch::simd::new
core::alloc::layout::{impl#0}::from_size_align_unchecked::precondition_check
core::alloc::layout::size
core::alloc::layout::align
core::alloc::layout::from_size_align_unchecked
core::alloc::layout::dangling
core::str::{impl#0}::as_bytes
core::slice::{impl#0}::as_ptr<u8>
core::cmp::impls::{impl#76}::lt
<unnamed>
alloc::string::from_utf8_unchecked
alloc::string::{impl#34}::to_string<str>
alloc::string::{impl#47}::from
alloc::string::{impl#111}::spec_to_string
alloc::vec::from_raw_parts_in<u8, alloc::alloc::Global>
alloc::vec::with_capacity_in<u8, alloc::alloc::Global>
alloc::vec::as_mut_ptr<u8, alloc::alloc::Global>
alloc::vec::set_len<u8, alloc::alloc::Global>
alloc::alloc::alloc_impl
alloc::alloc::alloc_zeroed
alloc::alloc::{impl#1}::allocate
alloc::alloc::{impl#1}::deallocate
alloc::alloc::exchange_malloc
alloc::alloc::alloc
alloc::alloc::dealloc
alloc::raw_vec::from_raw_parts_in<u8, alloc::alloc::Global>
alloc::raw_vec::with_capacity_in<u8, alloc::alloc::Global>
alloc::raw_vec::ptr<u8, alloc::alloc::Global>
alloc::raw_vec::from_raw_parts_in<alloc::alloc::Global>
alloc::raw_vec::needs_to_grow<alloc::alloc::Global>
alloc::raw_vec::capacity<alloc::alloc::Global>
alloc::raw_vec::with_capacity_in<alloc::alloc::Global>
alloc::raw_vec::non_null<alloc::alloc::Global, u8>
alloc::raw_vec::ptr<alloc::alloc::Global, u8>
alloc::raw_vec::new_cap<u8>
alloc::str::{impl#4}::to_owned
alloc::slice::{impl#0}::to_vec_in<u8, alloc::alloc::Global>
alloc::slice::{impl#0}::to_vec<u8>
alloc::slice::{impl#0}::into_vec<u8, alloc::alloc::Global>
alloc::slice::{impl#0}::to_vec_in::{impl#1}::to_vec<u8, alloc::alloc::Global>
alloc::slice::{impl#9}::to_owned<u8>
alloc::boxed::{impl#7}::into_raw_with_allocator<[u8], alloc::alloc::Global>
hashbrown::map::with_hasher<alloc::string::String, i32, std::hash::random::RandomState>
hashbrown::map::with_hasher<alloc::string::String, small::TestStruct0, std::hash::random::RandomState>
hashbrown::raw::new<(alloc::string::String, i32)>
hashbrown::raw::iter<(alloc::string::String, i32)>
hashbrown::raw::data_end<(alloc::string::String, i32)>
hashbrown::raw::buckets
hashbrown::raw::drop_elements<(alloc::string::String, i32)>
hashbrown::raw::iter<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::data_end<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::drop_elements<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::allocation_info
hashbrown::raw::free_buckets<alloc::alloc::Global>
hashbrown::raw::is_empty_singleton
hashbrown::raw::drop_inner_table<(alloc::string::String, i32), alloc::alloc::Global>
hashbrown::raw::drop_inner_table<(alloc::string::String, small::TestStruct0), alloc::alloc::Global>
hashbrown::raw::new<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::{impl#18}::drop<(alloc::string::String, i32), alloc::alloc::Global>
hashbrown::raw::{impl#18}::drop<(alloc::string::String, small::TestStruct0), alloc::alloc::Global>
hashbrown::raw::calculate_layout_for
hashbrown::raw::from_base_index<(alloc::string::String, i32)>
hashbrown::raw::as_ptr<(alloc::string::String, i32)>
hashbrown::raw::drop<(alloc::string::String, i32)>
hashbrown::raw::next_n<(alloc::string::String, i32)>
hashbrown::raw::new<(alloc::string::String, i32)>
hashbrown::raw::next_impl<(alloc::string::String, i32), false>
hashbrown::raw::{impl#29}::next<(alloc::string::String, i32)>
hashbrown::raw::{impl#29}::next<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::from_base_index<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::as_ptr<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::drop<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::next_n<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::new<(alloc::string::String, small::TestStruct0)>
hashbrown::raw::next_impl<(alloc::string::String, small::TestStruct0), false>
hashbrown::control::bitmask::lowest_set_bit
hashbrown::control::bitmask::nonzero_trailing_zeros
hashbrown::control::bitmask::remove_lowest_bit
hashbrown::control::bitmask::{impl#1}::into_iter
hashbrown::control::bitmask::{impl#1}::into_iter
hashbrown::control::bitmask::{impl#1}::into_iter
hashbrown::control::bitmask::{impl#1}::into_iter
hashbrown::control::bitmask::{impl#2}::next
hashbrown::control::bitmask::{impl#2}::next
hashbrown::control::group::neon::load_aligned
hashbrown::control::group::neon::match_full
hashbrown::util::invalid_mut<(alloc::string::String, i32)>
hashbrown::util::invalid_mut<(alloc::string::String, small::TestStruct0)>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
small::method_0
small::method_1
small::method_2
small::method_3
small::method_4
small::method_0
small::method_1
small::method_2
small::method_3
small::method_4
small::main
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
<unnamed>
